maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package io.opentelemetry.maven;
+
+import io.opentelemetry.api.trace.Span;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.apache.maven.model.Plugin;
+import org.apache.maven.plugin.MojoExecution;
+import org.apache.maven.project.MavenProject;
+import org.codehaus.plexus.component.annotations.Component;
+
+/** Holds the state of the spans in progress */
+@Component(role = SpanRegistry.class)
+public final class SpanRegistry {
+
+  private final Map<MojoExecutionKey, Span> mojoExecutionKeySpanMap = new HashMap<>();
+  private final Map<MavenProjectKey, Span> mavenProjectKeySpanMap = new HashMap<>();
+  private Span rootSpan;
+
+  @Nullable
+  public Span getRootSpan() {
+    return rootSpan;
+  }
+
+  /** @throws IllegalStateException Root span already defined */
+  public void setRootSpan(@Nonnull Span rootSpan) throws IllegalStateException {
+    if (this.rootSpan != null) {
+      throw new IllegalStateException("Root span already defined " + this.rootSpan);
+    }
+    this.rootSpan = rootSpan;
+  }
+
+  @Nonnull
+  public Span getSpan(@Nonnull MavenProject mavenProject) {
+    final MavenProjectKey key = MavenProjectKey.fromMavenProject(mavenProject);
+    final Span span = this.mavenProjectKeySpanMap.get(key);
+    if (span == null) {
+      throw new IllegalStateException(
+          "Span not started for project "
+              + mavenProject.getGroupId()
+              + ":"
+              + mavenProject.getArtifactId());
+    }
+    return span;
+  }
+
+  @Nonnull
+  public Span getRootSpanNotNull() {
+    if (rootSpan == null) {
+      throw new IllegalStateException("Root span not defined");
+    }
+    return rootSpan;
+  }
+
+  @Nonnull
+  public Span removeRootSpan() {
+    if (rootSpan == null) {
+      throw new IllegalStateException("Root span not defined");
+    }
+    if (!this.mojoExecutionKeySpanMap.isEmpty()) {
+      throw new IllegalStateException(
+          "Remaining children spans: "
+              + this.mojoExecutionKeySpanMap.keySet().stream()
+                  .map(MojoExecutionKey::toString)
+                  .collect(Collectors.joining(", ")));
+    }
+    return rootSpan;
+  }
+
+  public void putSpan(@Nonnull Span span, @Nonnull MavenProject mavenProject) {
+    MavenProjectKey key = MavenProjectKey.fromMavenProject(mavenProject);
+    Span previousSpanForKey = mavenProjectKeySpanMap.put(key, span);
+    if (previousSpanForKey != null) {
```

Comment:
What is being prevented with this check? Is there a situation that another span for the same `MavenProject` could be created?

Reply:
I have clarified with an IllegalStateException message. It should not happen, it would mean we have misunderstood the lifecycle of a Maven build.

Reply:
ok thanks